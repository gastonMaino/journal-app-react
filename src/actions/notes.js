import swal from 'sweetalert2';


import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
        dispatch(noteAdd(doc.id, newNote));
    }
}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

export const noteAdd = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);

        dispatch(setNotes(notes));
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url
        }

        const noteFirestore = { ...note };
        delete noteFirestore.id

        db.doc(`${uid}/journal/notes/${note.id}`).update(noteFirestore);

        dispatch(refreshNote(note.id, noteFirestore));

        swal.fire('Saved', note.title, 'success');
    }
}



export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        const { active } = getState().notes

        swal.fire({
            title: 'Uploading',
            text: 'Please wait...',
            allowOutsideClick: false,
            didOpen: () => {
                swal.showLoading();
            }
        })

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        swal.close();

        dispatch(startSaveNote(active))

    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));

        swal.fire('Completed', 'successfully note deleted', 'success')
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})