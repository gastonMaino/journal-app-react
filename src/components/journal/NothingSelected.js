import React from 'react'

export const NothingSelected = () => {
    return (
        <div className='nothing__main-content'>
            <p className='nothing__main-pargraph'>
                Select something
                <span className='nothing__main-paragraph-span'>or create an entry </span>
            </p>

            <i className='far fa-star fa-4x mt-3'></i>
        </div>
    )
}
