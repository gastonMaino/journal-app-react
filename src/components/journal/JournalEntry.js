import React from 'react'
import moment from 'moment'
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ id, date, title, body, url }) => {
    const dispatch = useDispatch()
    const noteDate = moment(date);

    const handleClick = () => {
        dispatch(activeNote(id, {
            date,
            title,
            body,
            url
        }))
    }

    return (
        <div className='journal__entry' onClick={handleClick} >
            {
                url &&
                <div
                    className='journal__entry-picture'
                    style={{
                        backgroundImage: `url(${url})`,
                        backgroundSize: 'cover',
                        backgroundPosition: 'center'
                    }
                    }
                ></div>
            }
            <div className='journal__entry-body'>
                <p className='journal__entry-title'>
                    {title}
                </p>
                <p className='journal__entry-content'>
                    { body }
                </p>
            </div>

            <div className='journal__entry-date-box'>
                <time>{noteDate.format('dddd')} {noteDate.format('Do')}</time>
            </div>
        </div >
    )
}
