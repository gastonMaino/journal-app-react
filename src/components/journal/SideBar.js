import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries';

export const SideBar = () => {

    const dispatch = useDispatch()

    const handleLogout = () =>{
        dispatch( startLogout() )
    }

    const { name } = useSelector(state => state.auth);

    const handleAddEntry = () =>{
        dispatch(startNewNote());
    }

    return (
        <aside className='journal__sidebar animate__animated animate__fadeIn animate__fast'>
            <div className='journal__sidebar-navbar'>
                <h1 className='mt-3 journal__sidebar-title'>
                    <i className='far fa-moon'></i>
                    <span>{name}</span>
                </h1>

                <button className='btn' onClick={handleLogout}>
                    logout
                </button>
            </div>

            <div className='journal__new-entry' onClick={handleAddEntry}>
                <i className='far fa-calendar-plus fa-5x'></i>
                <p className='mt-2'>new entry</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
