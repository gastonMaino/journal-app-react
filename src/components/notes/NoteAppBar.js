import React, { useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes'

export const NoteAppBar = () => {
    const dispatch = useDispatch()
    const { active } = useSelector(state => state.notes)

    const inputFile = useRef(null);

    const handleSave = () => {
        dispatch(startSaveNote(active))
    }

    const handlePictureClick = () =>{
        inputFile.current.click();
    }

    const handleFileChange = (e) => {
        const file = e.target.files[0];

        if( file ) {
            dispatch(startUploading(file))
        }
    }


    return (
        <div className='notes__appbar'>
            <time>June 12 2021</time>

            <input 
                type='file'
                style={{display: 'none'}}
                ref={inputFile}
                onChange={handleFileChange}
            />

            <div>
                <button
                    className='btn'
                    onClick={handlePictureClick}
                >
                    picture
                </button>
                <button
                    className='btn'
                    onClick={handleSave}
                >
                    save
                </button>
            </div>
        </div>
    )
}
