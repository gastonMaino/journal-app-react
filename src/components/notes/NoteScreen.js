import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';


import { activeNote, startDeleting } from '../../actions/notes';
import { NoteAppBar } from './NoteAppBar';
import { useForm } from '../../hooks/useForm/useForm';

export const NoteScreen = () => {
    const dispatch = useDispatch();

    const { active } = useSelector(state => state.notes)
    const [values, handleInputChange, reset] = useForm(active);
    const { title, body, id } = values;
    const activeId = useRef(active.id);

    useEffect(() => {
        if (active.id !== activeId.current) {
            reset(active);
            activeId.current = active.id
        }
    }, [active, reset])

    useEffect(() => {
        dispatch(activeNote(values.id, { ...values }));
    }, [values, dispatch])

    const handleDelete = () => {
        dispatch(startDeleting(id));
    }

    return (
        <div className='notes__main-content'>
            <NoteAppBar />

            <div className='notes__content'>
                <input
                    type='text'
                    placeholder='some awesome title'
                    className='notes__title-input'
                    name='title'
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder='what happened today?'
                    className='notes__textarea'
                    name='body'
                    value={body}
                    onChange={handleInputChange}
                ></textarea>
            </div>

            {
                (active.url) &&
                <div className='notes__image'>
                    <img src={active.url} alt='bird cage' />
                </div>
            }

            <button
                className='btn btn-danger'
                onClick={handleDelete}
            >
                delete
            </button>

        </div>
    )
}
