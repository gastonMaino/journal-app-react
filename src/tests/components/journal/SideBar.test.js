import { mount } from "enzyme/build"
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom'


import { SideBar } from "../../../components/journal/SideBar";
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";


jest.mock('../../../actions/auth', () =>({
    startLogout: jest.fn()
}))
jest.mock('../../../actions/notes', () =>({
    startNewNote: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {
        name: 'gaston'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes:{
        notes:[],
        active: null
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <SideBar />
        </MemoryRouter>
    </Provider>
);


describe('tes in SideBar', () => {
    
    test('should match to snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should call action startLogout', () => {
        wrapper.find('.btn').prop('onClick')();

        expect(startLogout).toHaveBeenCalled();
    })

    test('should call action ', () => {
        wrapper.find('.journal__new-entry').prop('onClick')()

        expect(startNewNote).toHaveBeenCalled();
    })
    
    
    

})
