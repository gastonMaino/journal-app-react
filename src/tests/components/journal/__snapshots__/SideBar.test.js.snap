// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tes in SideBar should match to snapshot 1`] = `
<aside
  className="journal__sidebar animate__animated animate__fadeIn animate__fast"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h1
      className="mt-3 journal__sidebar-title"
    >
      <i
        className="far fa-moon"
      />
      <span>
        gaston
      </span>
    </h1>
    <button
      className="btn"
      onClick={[Function]}
    >
      logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-2"
    >
      new entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
